dialog/file.go

```diff
@@ -480,20 +488,42 @@ func (f *fileDialog) setSelected(file fyne.URI, id int) {
 func (f *fileDialog) setView(view viewLayout) {
 	f.view = view
 	count := func() int {
+		f.dataLock.RLock()
+		defer f.dataLock.RUnlock()
+
 		return len(f.data)
 	}
 	template := func() fyne.CanvasObject {
 		return f.newFileItem(storage.NewFileURI("./tempfile"), true, false)
 	}
 	update := func(id widget.GridWrapItemID, o fyne.CanvasObject) {
+		f.dataLock.RLock()
+
+		if id >= len(f.data) {
+			f.dataLock.RUnlock()
+			return
+		}
+
 		dir := f.data[id]
+		f.dataLock.RUnlock()
+
 		parent := id == 0 && len(dir.Path()) < len(f.dir.Path())
 		_, isDir := dir.(fyne.ListableURI)
 		o.(*fileDialogItem).setLocation(dir, isDir || parent, parent)
 	}
 	choose := func(id int) {
+		f.dataLock.RLock()
```

Comment:
Perhaps this block of code that is repeated in the update and choose funcs could be extracted as its own helper:

```go
func (f *fileDialog) getDataItem(id int) (fyne.URI, bool) {
    f.dataLock.RLock()
    defer f.dataLock.RUnlock()
    if id >= len(f.data) {
        return nil, false
    }
    return f.data[id], true
}

Reply:
Great suggestion! Implemented.